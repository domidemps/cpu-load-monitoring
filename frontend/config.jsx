export const CONFIG = {
  pingInterval: 10, // seconds
  eventMinimumDuration: 120, // seconds
  historyWindow: 600, // seconds
}
