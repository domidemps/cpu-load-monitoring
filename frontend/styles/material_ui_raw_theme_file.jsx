import {createMuiTheme} from '@material-ui/core/styles'
import {fade} from '@material-ui/core'

const colors = {
  background: '#1f1f1f',
  paper: '#2f2f2f',
  text: '#e0e0e0',
}

const rawBaseTheme = {
  palette: {
    type: 'dark',
    background: {
      default: colors.background,
      paper: colors.paper,
    },
    text: {
      primary: colors.text,
      secondary: fade(colors.text, 0.7),
    },
  },
  overrides: {
    MuiAppBar: {
      colorPrimary: {
        color: colors.text,
        backgroundColor: colors.paper,
      },
    },
  },
}

export default createMuiTheme(rawBaseTheme)
