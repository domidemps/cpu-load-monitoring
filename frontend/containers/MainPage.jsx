/** @jsx jsx */
import {css, jsx} from '@emotion/core'
import AppBar from '@material-ui/core/AppBar'
import Toolbar from '@material-ui/core/Toolbar'
import Typography from '@material-ui/core/Typography'
import Percentage from './Percentage'
import Events from './Events'
import Plot from './Plot'
import {useEffect} from 'react'
import {getAverageLoad} from '../actions/load'
import {incrementTimer} from '../actions/timer'
import {useDispatch, useSelector} from 'react-redux'
import {addEvent, updateCurrentEvent, updatePreviousEvent} from '../actions/event'
import {getEvent, isEvent} from '../helpers/utils'
import {CONFIG} from '../config'

import last from 'lodash/last'
import isEmpty from 'lodash/isEmpty'

const styles = css`
  display: flex;
  flex-direction: column;
  .firstRow {
    display: flex;
  }
  .percentage {
    width: 40%;
  }
  .events {
    width: 60%;
  }
`

export default function MainPage() {
  const dispatch = useDispatch()
  const averageLoad = useSelector(state => state.load.averageLoad)
  const loadOverTime = useSelector(state => state.load.loadOverTime)
  const timer = useSelector(state => state.load.timer)
  const events = useSelector(state => state.load.events)

  // Ping the average CPU load every 10 seconds
  useEffect(() => {
    if (timer === 0) {
      dispatch(getAverageLoad())
    }
    const interval = setInterval(() => {
      dispatch(getAverageLoad())
      dispatch(incrementTimer())
    }, CONFIG.pingInterval * 1000)
    return () => clearInterval(interval)
  }, [averageLoad])

  // Heavy CPU load:
  // - average CPU load > 1
  // - since more than 2 minutes
  // - no events yet or previous event is a recovery
  useEffect(() => {
    if (isEvent(loadOverTime, 'heavy') && (isEmpty(events) || last(events).type === 'recovery')) {
      const newEventStart = getEvent(loadOverTime, CONFIG.eventMinimumDuration).time
      const newEvent = {
        type: 'heavy',
        start: newEventStart,
        end: null,
      }
      dispatch(addEvent(newEvent))
      dispatch(updateCurrentEvent(newEvent))
      // Update end time of the previous event
      if (last(events)?.type === 'recovery') {
        dispatch(updatePreviousEvent(newEvent, {end: newEventStart}))
      }
    }
    // Recovery of heavy CPU load:
    // - average CPU load < 1
    // - since more than 2 minutes
    // - previous event is a heavy CPU load
    else if (
      isEvent(loadOverTime, 'recovery') &&
      !isEmpty(events) &&
      last(events).type === 'heavy'
    ) {
      const newEventStart = getEvent(loadOverTime, CONFIG.eventMinimumDuration).time
      const newEvent = {
        type: 'recovery',
        start: newEventStart,
        end: null,
      }
      dispatch(addEvent(newEvent))
      dispatch(updateCurrentEvent(newEvent))
      dispatch(updatePreviousEvent(newEvent, {end: newEventStart}))
    }
  })

  return (
    <div css={styles}>
      <AppBar position="static">
        <Toolbar>
          <Typography variant="h5">CPU load monitoring</Typography>
        </Toolbar>
      </AppBar>
      <div className="firstRow">
        <Percentage className="percentage" />
        <Events className="events" events={events} />
      </div>
      <Plot />
    </div>
  )
}
