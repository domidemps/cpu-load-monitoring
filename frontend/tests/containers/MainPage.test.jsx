import React from 'react'
import configureStore from 'redux-mock-store'
import thunk from 'redux-thunk'
import {Provider} from 'react-redux'
import renderer from 'react-test-renderer'

import MainPage from '../../containers/MainPage'
import {CONFIG} from '../../config'
import {COLORS} from '../../styles/material_ui_raw_theme_file'

describe('MainPage', () => {
  let store

  it('renders correctly', () => {
    store = configureStore([thunk])({
      load: {
        averageLoad: 0.5,
        loadOverTime: new Array(CONFIG.historyWindow / CONFIG.pingInterval).fill().map((_, i) => ({
          load: 50,
          time: '12:00:00',
          color: COLORS.normalLoad,
        })),
        timer: 10,
        events: [{type: 'heavy', start: '12:30:00', end: '12:33:00'}],
        currentEvent: {type: 'recovery', start: '12:33:10', end: null},
      },
    })
    expect(
      renderer
        .create(
          <Provider store={store}>
            <MainPage />
          </Provider>,
        )
        .toJSON(),
    ).toMatchSnapshot()
  })
})
