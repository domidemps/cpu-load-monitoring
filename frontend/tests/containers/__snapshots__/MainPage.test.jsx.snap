// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MainPage renders correctly 1`] = `
<div
  className="css-iwvqqw-MainPage"
>
  <header
    className="MuiPaper-root MuiAppBar-root MuiAppBar-positionStatic MuiAppBar-colorPrimary MuiPaper-elevation4"
  >
    <div
      className="MuiToolbar-root MuiToolbar-regular MuiToolbar-gutters"
    >
      <h5
        className="MuiTypography-root MuiTypography-h5"
      >
        CPU load monitoring
      </h5>
    </div>
  </header>
  <div
    className="firstRow"
  >
    <div
      className="MuiPaper-root css-r93fnb-Percentage MuiPaper-elevation5 MuiPaper-rounded"
    >
      <div
        className="css-qy1xbs-Percentage"
      >
        <h5
          className="MuiTypography-root css-1gri5ec-Percentage MuiTypography-h5"
        >
          Average CPU load
        </h5>
        <h1
          className="MuiTypography-root css-1h9ee3p-Percentage MuiTypography-h1"
        >
          50%
        </h1>
      </div>
    </div>
    <div
      className="MuiPaper-root css-h168i7-Events MuiPaper-elevation5 MuiPaper-rounded"
    >
      <div
        className="alignCenter"
      >
        <h5
          className="MuiTypography-root MuiTypography-h5"
        >
          Events
        </h5>
        <div
          className="bigSpaceAround"
        >
          <span
            className="heavyLoad"
          >
            1
             heavy CPU loads, 
          </span>
          <span
            className="normalLoad"
          >
            0
             recovered
          </span>
        </div>
        <div
          className="scrollable"
        >
          <div
            className="MuiPaper-root MuiAlert-root MuiAlert-standardError spaceAround MuiPaper-elevation0"
            role="alert"
          >
            <div
              className="MuiAlert-icon"
            >
              <svg
                aria-hidden={true}
                className="MuiSvgIcon-root MuiSvgIcon-fontSizeInherit"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <path
                  d="M11 15h2v2h-2zm0-8h2v6h-2zm.99-5C6.47 2 2 6.48 2 12s4.47 10 9.99 10C17.52 22 22 17.52 22 12S17.52 2 11.99 2zM12 20c-4.42 0-8-3.58-8-8s3.58-8 8-8 8 3.58 8 8-3.58 8-8 8z"
                />
              </svg>
            </div>
            <div
              className="MuiAlert-message"
            >
              System under heavy CPU load from 12:30:00 to 12:33:00
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
  <div
    className="MuiPaper-root css-jfobc3-Plot MuiPaper-elevation5 MuiPaper-rounded"
  >
    <div
      style={
        Object {
          "height": 550,
          "width": "100%",
        }
      }
    >
      <h5
        className="MuiTypography-root css-4wzwv1-Plot MuiTypography-h5"
      >
        Average CPU load over time
      </h5>
      <div
        style={
          Object {
            "height": "100%",
            "width": "100%",
          }
        }
      />
    </div>
  </div>
</div>
`;
